mansrc/fastBlockSearch.1
'\" t
.\"     Title: fastblocksearch
.\"    Author: [see the "AUTHORS" section]
.\" Generator: Asciidoctor 1.5.5
.\"      Date: 
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "FASTBLOCKSEARCH" "1" "" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.LINKSTYLE blue R < >
.SH "NAME"
fastblocksearch \- search loci matching protein block profiles
.SH "SYNOPSIS"
.sp
\fBfastBlockSearch\fP [options] seqs.fa fam.prfl
.SH "DESCRIPTION"
.sp
.if n \{\
.RS 4
.\}
.nf
Searches hits (matches) of the blocks in the profile given by fam.prfl within the genomic sequences in the file seqs.fa.
Hits are sorted increasingly by score, so the last displayed hit is the best one found in the region. The format is similar
to that of the blockscore file (which is optionally generated by msa2prfl.pl): It shows coordinate, strand,
mean odds\-ratio score, and specificity of score, and the motif.
From the output users can chose regions with matching blocks to perform gene prediction with AUGUSTUS\-PPX using the same block profile.
.fi
.if n \{\
.RE
.\}
.SH "EXAMPLE"
.sp
.if n \{\
.RS 4
.\}
.nf
> fastBlockSearch \-\-cutoff=1.1 chr4.103M.fa PF00225_seed.prfl
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
Hits found in chr4 103000000 105000000
Score:207.987
Mult. score:4.83391
1081586 unknown_M[5,13] \-       2.32574 5.04633 .....YATRLKNI
1103952 unknown_L       \-       4.85363 6.75245 NAKTRIICTITP
1103991 unknown_K       \-       8.38065 9.92928 YRDSKLTRILQNSLG
1104375 unknown_J       \-       3.96065 6.79408 RSLFILGQVIKKL
1106992 unknown_I       \-       9.22487 7.64306 LVDLAGSE
1115567 unknown_H[5,16] \-       2.31869 5.58986 .....ESRHYGETKMN
1116319 unknown_G       \-       7.34282 8.29425 EIYNETITDLL
1117092 unknown_F       \-       5.10694 6.10274 VIPRAIHDIF
1117146 unknown_E       \-       9.43596 9.18891 QTASGKTYTM
1117176 unknown_D[1,8]  \-       5.73796 6.31532 .GTIFAYG
1117399 unknown_B[1,7]  \-       3.59083 5.03059 .CLDRVF
1119420 unknown_A[0,8]  \-       4.64107 6.44285 RVRPLNSR.
.fi
.if n \{\
.RE
.\}
.SH "OPTIONS"
.sp
\fB\-\-cutoff=c\fP
.RS 4
This minimum for the average log score of the motifs found can be used to adjust the sensitivity of the block search.
The standard cutoff is 0.7, which is very sensitive but can give many false positive hits for smaller profiles.
.RE
.SH "AUTHORS"
.sp
Oliver Keller

HISTORY.TXT
     - remove bam2wig
     - compile with CGP and ZIPINPUT support by default
     - compile with SQLite and MySQL support for CGP (switch of by adding SQLite = false and MySQL = false to common.mk)
List of changes from version 3.3.2 to 3.3.3 (until Sep 13th, 2019)
     - docker container
     - new script tfix_in_frame_stop_codon_genes.py that replaces genes where spliced in-frame stop codons were predicted
     - new scripts: compare_masking.pl, merge_masking.pl
     - new program bamToWig.py as alternarive to bam2wig C++ binary
     - fix warnings on new GCC compiler (8.3)
     - introduced unit tests of C++ code (make test)
     - bugfixes (lex.cc, autoAug.pl)
     - DIAMOND as alternative to BLAST in aa2nonred.pl
List of changes from version 3.3.1 to 3.3.2 (until Oct 5th, 2018)
     - bugfixes in comparative augustus, utrrnaseq
     - new species Chiloscyllium punctatum (bamboo shark), Scyliorhinus torazame (cat shark), Rhincodon typus (whale shark)
     - updated comparative augustus (CGP) tutorial 
List of changes from version 3.3 to 3.3.1 (until May 8th, 2018)
     - new species pisaster (Pisaster ochraceus, ochre starfish)
     - bugfixes of sampling error in intron model and in check of
       frame-compatibility of CDSpart hints without frame
     - new auxiliary program utrrnaseq to find training UTRs
     - new script setStopCodonFreqs.pl
List of changes from version 3.2.3 to 3.3 (until July 11th, 2017)
     - new program ESPOCA to estimate selective pressure on codon alignments
     - gene finding on ancestral genomes is enabled
     - new default parameters for comparative gene prediction (CGP)
     - clade parameters training for CGP
     - compatibility to Ian Fiddes' Comparative Annotation Toolkit (CAT)
     - new scripts eval_dualdecomp.pl,
     - more tolerant tree parsing
     - bugfixes in augustus, joingenes, load2sqlitedb, transMap2hints.pl, splitMfasta.pl, intron2exex.pl,
       aln2wig
     - new functionality in homGeneMapping, joingenes
     - new species ciona, mnemiopsis_leidyi, nematostella_vectensis, strongylocentrotus_purpuratus
List of changes from version 3.2.2 to 3.2.3 (until December 6th, 2016)
     - bugfixes in joingenes, load2sqlitedb, transMap2hints.pl
     - added script gtf2bed.pl
     - CGP parameters for clade of 8 primates (cgp/log_reg_parameters_primates.cfg)
     - added option --clean to load2sqlitedb
     - now filter out truncated coding genes with CDS length <= 3
     - Sascha Steinbiss: remove dependency on PATH_MAX
List of changes from version 3.2.1 to 3.2.2 (until March 30th, 2016)
     - added extrinsic.M.RM.PB.cfg for PacBio CCS hints with GMAP
       (see http://bioinf.uni-greifswald.de/bioinf/wiki/pmwiki.php?n=Augustus.PacBioGMAP )
     - standardization of Makefile
     - improvements in auxiliary program joingenes
     - added missing program load2sqlitedb
     - comparative gene predictions (CGP):
         - hints with source=M (e.g. existing annotations) are fixed
	 - new standard parameters for exon_gain, exon_loss and others
     - bugfixes, in particular a double free or corruption bug in PPX that occured with BUSCO
     - new species Volvox carteri (volvox)
List of changes from version 3.2 to 3.2.1 (until October 23rd, 2015)
     - comparative gene prediction (CGP): make the parameters from the GCB proceedings paper match again
     - new CGP default parameters (tested on 5 clades)
     - CGP version penalizes very long introns equivalently to a geometric distribution
     - code now under artistic license only
List of changes from version 3.1 to 3.2 (until September 25th, 2015)
     - comparative gene prediction (CGP option) has sensible default parameters of the logistic regression model
     - new tool hal2maf_split.pl for splitting an alignment for parallel comparative prediction
     - bugfixes to augustus (compiler warnings on MaxOS)
     - improvements to prediction of noncoding genes
     - homGeneMapping can determine homologous sets of transcripts
     - bugfixes in joingenes and bam2hints (rare segmentation fault)
     - made load2sqlitedb more robust, new option to skip index building when loading many species
     - new tutorial that includes iterative training
     - new option to shift coordinates in hint input when sequence window is cut out
     - adapted cegma2gff.pl to cegma version 2.5
List of changes from version 3.0.3 to 3.1 (until April 29th, 2015)
     - prediction of noncoding genes (ncRNA) with option --nc=on, requires input of hints, e.g. from RNA-Seq,
       predicts coding genes (with UTRs) and noncoding genes simultaneously
     - speed improvement with --UTR=1 (by 83% on human without sampling, by 18% on fly with sampling)
     - softmasking support: --softmasking=1 is a little more accurate and much faster on human
     - small overall accuracy improvement
     - prevent the prediction of spliced in-frame stop codons (for SHORT or HINTED introns only)
     - new species: rice, Danio rerio (zebrafish), s_aureus (Staphylococcus aureus),
       camponotus_floridanus (ant) contributed by Shishir K Gupta,
       sulfolobus_solfataricus (archaeon), Parasteatoda tepidariorum (house spider), Apis dorsata (from Francisco Camara Ferreira)
     - update to parameters of Thermoanaerobacter tencongensis, E.coli, Yarrowia lipolytica, schistosoma2 (Schistosoma mansoni)
     - remove dependency on libboost-filesystems, boost is now completely optional in single-species mode
     - new auxiliary program 'homGeneMapping' for identification of homologous genes in a set of gene predictions
     - new auxiliary program 'joingenes' for combining gene sets
     - bugfixes in load2db, length distribution skew of bacteria, comparative mode (findGeneRanges,
       Alignment::mergeable, memory leak), some ass hints were previously not used although possible
     - decreased minintronlen default of bam2hints to 32 as the old threshold tended to be too large
     - computing of omega = dN/dS more efficiently

src/fastBlockSearch.cc
/*
 * fastBlockSearch.cc
 *
 * License: Artistic License, see file LICENSE.TXT or 
 *          https://opensource.org/licenses/artistic-license-1.0
 */

// #include <iostream>
// #include <vector>
// #include <fstream>
// #include <iomanip>
// #include <sstream>
// #include <string>
// #include <set>
#include <iostream>
#include <fstream>
#include <iomanip>

#define VERBOSE_SEEDS 1

#include "properties.hh"
#include "pp_profile.hh"
#include "pp_fastBlockSearcher.hh"

// #define SEEDSIZE 6
// #define SEED_COLUMNSCORE 0.8
#define MAX_SEQSIZE (unsigned(-1)/2)

//#define TESTLEN 1000000
#define MINCOUNT 0

using namespace std;

// set this with --cutoff=x
// or with --avscore=exp(x)
double cutoff = 0.7; //
   
// struct IndexComp {
//     IndexComp(const PP::Column& v) : vec(v) {}
//     const PP::Column& vec;
//     bool operator() (int a, int b) {
// 	return vec.Q(a) < vec.Q(b);
//     }
// };

// struct PrefilterType {
//     PrefilterType() : seedhitcount(0), lastOffset(-10) {}
//     int seedhitcount;
//     int lastOffset;
// };


void printBestResults(PP::Profile& prfl, PP::FsHitCollection& searchResults, double cutoff, double columnCount, int offset) {
    multimap<double, PP::FsHitType> bestScores;
    searchResults.storeBestResults(bestScores, MINCOUNT, cutoff * columnCount);
    for (map<double,PP::FsHitType>::iterator it=bestScores.begin();
	 it != bestScores.end();
	 ++it) {
	vector<string> output(1, "--\n");
	cout << "Score:" << it->first << endl;
	cout << "Mult. score:" << exp(it->first / columnCount) << endl;
	for (PP::FsHitType* ht = &(it->second); ht != 0; ht = ht->predecessor) {
	    int startpos = ht->pos + offset + 
		3 * (ht->reverse ? 
		     prfl.blockSize(ht->blockNo)-ht->blockto :
		     ht->blockfrom);

	    ostringstream sstrm;

	    sstrm << startpos << "\t" << prfl[ht->blockNo].id;
	    if (ht->blockfrom != 0 || ht->blockto != prfl.blockSize(ht->blockNo)) 
		sstrm << "[" << ht->blockfrom << "," << ht->blockto << "]";
	    sstrm << "\t" << (ht->reverse ? '-' : '+') << "\t" 
		  << exp(ht->score / prfl.blockSize(ht->blockNo)) << "\t" 
//		  << ht->score << "\t" 
		  << prfl[ht->blockNo].getBackDist().normed(ht->score);
	    string blseq;
	    for (int i=0; i < prfl.blockSize(ht->blockNo); i++) {
		blseq += (i < ht->blockfrom || i >= ht->blockto) ? '.' : ht->reverse ? 
		    GeneticCode::revtranslate(PP::DNA::sequence + ht->pos + 3*(prfl.blockSize(ht->blockNo) -1 -i)) :
		    GeneticCode::translate(PP::DNA::sequence + ht->pos + 3*i);
	    }
	    sstrm << "\t" << blseq; // string(PP::DNA::sequence + ht->pos, prfl.blockSize(ht->blockNo)*3); 
	    sstrm << "\n";
#ifdef DEBUG
	    sstrm << string(PP::DNA::sequence + startpos - offset, 3*(ht->blockto - ht->blockfrom)) << endl;
#endif
	    output.push_back(sstrm.str());
	}
	while (!output.empty()) {
	    cout << output.back();
	    output.pop_back();
	}
    }
}



int main(int argc, char* argv[]) {
    bool abort = false;
    vector<string> argvec;
    for (int i=1; i<argc; i++) {
	string param(argv[i]);
	if (param.substr(0,2) == "--") {
	    int i = param.find("=");
	    if (param.find("=") != string::npos) {
		string key = param.substr(2, i-2);
		string value = param.substr(i+1);
		istringstream istrm(value);
		if (key == "cutoff" || key == "avscore") {
		    abort = !(istrm >> cutoff);
		    if (key == "avscore") 
			cutoff = log(cutoff);
		} else
		    Properties::addProperty(key, value);
	    } else
		abort = true;
	} else
	    argvec.push_back(param);
	if (abort) break;
    }

    if (abort || argvec.size() != 2) {
	cerr << "Usage: fastBlocksSearch [options] <seqs.fa> <fam.prfl>\n";
	return -1;
    }

    // read in sequence file
//     map<string, SeqType> sequences;
    ifstream fstrm(argvec[0].c_str());
    if (!fstrm) {
	cerr << "Could not open \"" << argvec[0] << "\".\n";
	return -1;
    }
    
// Properties::addProperty("/ProteinModel/block_threshold_spec", "0.0");
    // Properties::addProperty("/ProteinModel/block_threshold_sens", "0.0");
    PP::initConstants();
    // PP::Profile::min_anchor_count = 0;

    // read in ptn
    PP::Profile prfl(argvec[1].c_str());
    
    int columnCount = 0;
    PP::FsSeedCollection seedColl(prfl);
    for (int b=0; b<prfl.blockCount(); b++)
	columnCount += prfl.blockSize(b);

    cerr << "Profile has " << prfl.blockCount() << " blocks with " << columnCount << " columns." << endl;
    cerr << "Total seed count: " << seedColl.size() << endl;
    cerr << "Reading sequence(s)";
    int predictionStart=0, predictionLen=MAX_SEQSIZE;
    try {
	predictionStart = Properties::getIntProperty("predictionStart") -1;
    } catch (...) {}
    try {
	predictionLen = Properties::getIntProperty("predictionEnd") - predictionStart;
    } catch (...) {}
    if (predictionLen <= 0) {
	cerr << "predictionStart must be less than predictionEnd!" << endl;
	exit(-1);
    }
    while (fstrm) {
	PP::FsHitCollection fastSearchResults(prfl.blockCount());

	string seqname, currSeq;
	getline(fstrm, seqname);
	if (seqname[0] != '>') {
	    cerr << "Not in FASTA format.\n";
	    return -1;
	}
	// int pos=seqname.find_first_of("\t\n\v\f\r ");
	// if (pos>=0)
	//     seqname.erase(pos);
	seqname = seqname.substr(1);
	
	int readlen = 0;
	int progress = 0;
	while (fstrm && fstrm.peek() != '>') {
	    string line;
	    getline(fstrm, line);
	    int pos = 0;
	    for (int i=0; i<line.length() && pos < predictionLen - currSeq.length(); i++) 
		if (isalpha(line[i])) //  || currSeq[i] == '-' || currSeq[i] == '.')
		    line[pos++] = tolower(line[i]);
	    line.erase(pos);
	    pos += readlen;
	    if (predictionStart <= pos) {
		if (predictionStart > readlen) 
		    currSeq += line.substr(predictionStart - readlen);
		else  
		    currSeq += line;
	    }
	    readlen = pos;
	    int newprogress = readlen/1000000;
	    while (newprogress > progress) {
		progress++;
		if (progress%10) 
		    cerr << ".";
		else
		    cerr << " " << progress << "M ";
	    }
	    if (currSeq.length() >= predictionLen) {
		if (currSeq.length() > predictionLen) 
		    cerr << "ERROR: prediction Length incorrect!" << endl;
		break;
	    }
	}
	cerr << endl << "Read " << readlen << " bps from '" << seqname << "'.";
	if (predictionLen < MAX_SEQSIZE) 
	    cerr << " Using " << predictionLen << ".";
	cerr << endl;

	PP::DNA::initSeq(currSeq);
	PP::CandidateCollection candidates(prfl, seedColl);
	progress = 0;
	for (int t=2; t<currSeq.length(); t++) {
	    candidates.proceed(t, fastSearchResults);

	    double newprogress = (100.0 * (t+1)) / currSeq.length() -0.1;
	    while (progress < newprogress) {
		progress++;
		switch (progress%10) {
		    case 0 : cerr << progress; break;
		    case 9 : break;
		    case 8 : if (progress != 98) cerr << ".";
		    default: cerr << ".";
		}
	    }
	}
	cerr << "\n # hits: " << fastSearchResults.allHitCount() << endl; 
	cerr << " # hit groups: " << fastSearchResults.resultCount() << endl;
	cerr << "output will show all hits above " << cutoff * columnCount;
	if (MINCOUNT > 0)
	    cerr << ", at least " << MINCOUNT;
	cerr << ".\n";
    
	cout << "Hits found in " << seqname << endl;
	printBestResults(prfl, fastSearchResults, cutoff, columnCount, predictionStart);
	cout << endl;
        delete [] PP::DNA::sequence;
        
	if (predictionStart > 0 || predictionLen != MAX_SEQSIZE)
	    break;
    }
}
